File: src/CVNBot/Project.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;
using System.Xml;
using log4net;

namespace CVNBot
{
    class Project
    {
        static ILog logger = LogManager.GetLogger("CVNBot.Project");

        public string projectName;
        public string interwikiLink;
        public string rooturl; // Format: https://en.wikipedia.org/

        public Regex rrestoreRegex;
        public Regex rdeleteRegex;
        public Regex rprotectRegex;
        public Regex runprotectRegex;
        public Regex rmodifyprotectRegex;
        public Regex ruploadRegex;
        public Regex rmoveRegex;
        public Regex rmoveredirRegex;
        public Regex rblockRegex;
        public Regex runblockRegex;
        public Regex rreblockRegex;
        public Regex rautosummBlank;
        public Regex rautosummReplace;
        public Regex rSpecialLogRegex;
        public Regex rCreate2Regex;

        public Hashtable namespaces;

        Dictionary<string, string> regexDict = new Dictionary<string, string>();

        static char[] rechars = {'\\', '.' ,'(', ')', '[' , ']' ,'^' ,'*' ,'+' ,'?' ,'{' ,'}' ,'|' };
        string snamespaces;

        /// <summary>
        /// Generates Regex objects from regex strings in class. Always generate the namespace list before calling this!
        /// </summary>
        void GenerateRegexen()
        {
            rrestoreRegex = new Regex(regexDict["restoreRegex"]);
            rdeleteRegex = new Regex(regexDict["deleteRegex"]);
            rprotectRegex = new Regex(regexDict["protectRegex"]);
            runprotectRegex = new Regex(regexDict["unprotectRegex"]);

            if (!regexDict.ContainsKey("modifyprotectRegex"))
            {
                // Added in CVNBot 1.20, fallback if missing in older XML files.
                regexDict["modifyprotectRegex"] = regexDict["protectRegex"];
                logger.Warn("generateRegexen: modifyprotectRegex is missing. Please reload this wiki.");
            }
            rmodifyprotectRegex = new Regex(regexDict["modifyprotectRegex"]);
            ruploadRegex = new Regex(regexDict["uploadRegex"]);
            rmoveRegex = new Regex(regexDict["moveRegex"]);
            rmoveredirRegex = new Regex(regexDict["moveredirRegex"]);
            rblockRegex = new Regex(regexDict["blockRegex"]);
            runblockRegex = new Regex(regexDict["unblockRegex"]);
            if (!regexDict.ContainsKey("reblockRegex")) {
                // Added in CVNBot 1.22, fallback if missing in older XML files.
                regexDict["reblockRegex"] = "^$";
                logger.Warn("generateRegexen: reblockRegex is missing. Please reload this wiki.");
            }
            rreblockRegex = new Regex(regexDict["reblockRegex"]);
            rautosummBlank = new Regex(regexDict["autosummBlank"]);
            rautosummReplace = new Regex(regexDict["autosummReplace"]);

            rSpecialLogRegex = new Regex(regexDict["specialLogRegex"]);

            rCreate2Regex = new Regex( namespaces["2"]+@":([^:]+)" );
        }

        public string DumpProjectDetails()
        {
            StringWriter output = new StringWriter();

            using (XmlTextWriter dump = new XmlTextWriter(output))
            {
                dump.WriteStartElement("project");

                dump.WriteElementString("projectName", projectName);
                dump.WriteElementString("interwikiLink", interwikiLink);
                dump.WriteElementString("rooturl", rooturl);
                dump.WriteElementString("speciallog", regexDict["specialLogRegex"]);
                dump.WriteElementString("namespaces", snamespaces.Replace("<?xml version=\"1.0\" encoding=\"utf-8\"?>", ""));

                dump.WriteElementString("restoreRegex", regexDict["restoreRegex"]);
                dump.WriteElementString("deleteRegex", regexDict["deleteRegex"]);
                dump.WriteElementString("protectRegex", regexDict["protectRegex"]);
                dump.WriteElementString("unprotectRegex", regexDict["unprotectRegex"]);
                dump.WriteElementString("modifyprotectRegex", regexDict["modifyprotectRegex"]);
                dump.WriteElementString("uploadRegex", regexDict["uploadRegex"]);
                dump.WriteElementString("moveRegex", regexDict["moveRegex"]);
                dump.WriteElementString("moveredirRegex", regexDict["moveredirRegex"]);
                dump.WriteElementString("blockRegex", regexDict["blockRegex"]);
                dump.WriteElementString("unblockRegex", regexDict["unblockRegex"]);
                dump.WriteElementString("reblockRegex", regexDict["reblockRegex"]);
                dump.WriteElementString("autosummBlank", regexDict["autosummBlank"]);
                dump.WriteElementString("autosummReplace", regexDict["autosummReplace"]);

                dump.WriteEndElement();
                dump.Flush();
            }

            return output.ToString();
        }

        public void ReadProjectDetails(string xml)
        {
            XmlDocument doc = new XmlDocument();
            doc.LoadXml(xml);
            XmlNode parentnode = doc.FirstChild;
            for (int i = 0; i < parentnode.ChildNodes.Count; i++)
            {
                string key = parentnode.ChildNodes[i].Name;
                string value = parentnode.ChildNodes[i].InnerText;
                switch (key)
                {
                    case "projectName": projectName = value; break;
                    case "interwikiLink": interwikiLink = value; break;
                    case "rooturl": rooturl = value; break;
                    case "speciallog": regexDict["specialLogRegex"] = value; break;
                    case "namespaces": snamespaces = value; break;
                    case "restoreRegex": regexDict["restoreRegex"] = value; break;
                    case "deleteRegex": regexDict["deleteRegex"] = value; break;
                    case "protectRegex": regexDict["protectRegex"] = value; break;
                    case "unprotectRegex": regexDict["unprotectRegex"] = value; break;
                    case "modifyprotectRegex": regexDict["modifyprotectRegex"] = value; break;
                    case "uploadRegex": regexDict["uploadRegex"] = value; break;
                    case "moveRegex": regexDict["moveRegex"] = value; break;
                    case "moveredirRegex": regexDict["moveredirRegex"] = value; break;
                    case "blockRegex": regexDict["blockRegex"] = value; break;
                    case "unblockRegex": regexDict["unblockRegex"] = value; break;
                    case "reblockRegex": regexDict["reblockRegex"] = value; break;
                    case "autosummBlank": regexDict["autosummBlank"] = value; break;
                    case "autosummReplace": regexDict["autosummReplace"] = value; break;
                }
            }
            // Always get namespaces before generating regexen
            GetNamespaces(true);
            // Regenerate regexen
            GenerateRegexen();
        }

        void GetNamespaces(bool snamespacesAlreadySet)
        {
            if (!snamespacesAlreadySet)
            {
                logger.InfoFormat("Fetching namespaces from {0}", rooturl);
                snamespaces = CVNBotUtils.GetRawDocument(rooturl + "w/api.php?format=xml&action=query&meta=siteinfo&siprop=namespaces");
                if (snamespaces == "")
                    throw new Exception("Can't load list of namespaces from " + rooturl);
            }

            namespaces = new Hashtable();

            XmlDocument doc = new XmlDocument();
            doc.LoadXml(snamespaces);
            string namespacesLogline = "";
            XmlNode namespacesNode = doc.GetElementsByTagName("namespaces")[0];
            for (int i = 0; i < namespacesNode.ChildNodes.Count; i++)
            {
                namespaces.Add(namespacesNode.ChildNodes[i].Attributes["id"].Value, namespacesNode.ChildNodes[i].InnerText);
                namespacesLogline += "id["+namespacesNode.ChildNodes[i].Attributes["id"].Value + "]="+namespacesNode.ChildNodes[i].InnerText + "; ";
            }
        }

        public struct MessagesOption
        {
            public int NumberOfArgs;
            public string RegexName;
            public bool NonStrictFlag;
            public MessagesOption (int ArgNumberOfArgs, string ArgRegexName, bool ArgNonStrictFlag)
            {
                NumberOfArgs = ArgNumberOfArgs;
                RegexName = ArgRegexName;
                NonStrictFlag = ArgNonStrictFlag;
            }
        }

        public void RetrieveWikiDetails()
        {
            //Find out what the localized Special: (ID -1) namespace is, and create a regex
            GetNamespaces(false);

            regexDict["specialLogRegex"] = namespaces["-1"] + @":.+?/(.+)";

            logger.InfoFormat("Fetching interface messages from {0}", rooturl);

            Dictionary<string, MessagesOption> Messages = new Dictionary<string, MessagesOption>();

            // Location of message, number of required parameters, reference to regex, allow lazy
            // Retrieve messages for all the required events and generate regexen for them

            Messages.Add("Undeletedarticle", new MessagesOption(1, "restoreRegex", false));
            Messages.Add("Deletedarticle", new MessagesOption(1, "deleteRegex", false));
            Messages.Add("Protectedarticle", new MessagesOption(1, "protectRegex", false));
            Messages.Add("Unprotectedarticle", new MessagesOption(1, "unprotectRegex", false));
            Messages.Add("Modifiedarticleprotection", new MessagesOption(1, "modifyprotectRegex", true));
            Messages.Add("Uploadedimage", new MessagesOption(0, "uploadRegex", false));
            Messages.Add("1movedto2",new MessagesOption(2, "moveRegex", false));
            Messages.Add("1movedto2_redir", new MessagesOption(2, "moveredirRegex", false));

            // blockRegex is nonStrict because some wikis override the message without including $2 (block length).
            // RCReader will fall back to "24 hours" if this is the case.
            // Some newer messages (e.g. https://lmo.wikipedia.org/wiki/MediaWiki:Blocklogentry) have a third item,
            // $3 ("anononly,nocreate,autoblock"). This may conflict with $2 detection.
            // Trying (changed 2 -> 3) to see if length of time will be correctly detected using just this method:
            Messages.Add("Blocklogentry", new MessagesOption(3, "blockRegex", true));

            Messages.Add("Unblocklogentry", new MessagesOption(0, "unblockRegex", false));
            Messages.Add("Reblock-logentry", new MessagesOption(3, "reblockRegex", false));
            Messages.Add("Autosumm-blank", new MessagesOption(0, "autosummBlank", false));

            // autosummReplace is nonStrict because some wikis use translations overrides without
            // a "$1" parameter for the content.
            Messages.Add("Autosumm-replace", new MessagesOption(1, "autosummReplace", true));

            GetInterfaceMessages(Messages);

            GenerateRegexen();
        }

        void GetInterfaceMessages(Dictionary<string, MessagesOption> Messages)
        {
            string CombinedMessages = string.Join("|", Messages.Keys);

            string sMwMessages = CVNBotUtils.GetRawDocument(
                rooturl +
                "w/api.php?action=query&meta=allmessages&format=xml" +
                "&ammessages=" + CombinedMessages
            );
            if (sMwMessages == "")
                throw new Exception("Can't load list of InterfaceMessages from " + rooturl);

            XmlDocument doc = new XmlDocument();
            doc.LoadXml(sMwMessages);
            string mwMessagesLogline = "";
            XmlNode allmessagesNode = doc.GetElementsByTagName("allmessages")[0];
            for (int i = 0; i < allmessagesNode.ChildNodes.Count; i++)
            {
                string elmName = allmessagesNode.ChildNodes[i].Attributes["name"].Value;
                GenerateRegex(
                    elmName,
                    allmessagesNode.ChildNodes[i].InnerText,
                    Messages[elmName].NumberOfArgs,
                    Messages[elmName].RegexName,
                    Messages[elmName].NonStrictFlag
                );
                mwMessagesLogline += "name[" + elmName + "]="+allmessagesNode.ChildNodes[i].InnerText + "; ";
            }
        }

        void GenerateRegex(string mwMessageTitle, string mwMessage, int reqCount, string destRegex, bool nonStrict)
        {
            // Now gently coax that into a regex
            foreach (char c in rechars)
                mwMessage = mwMessage.Replace(c.ToString(), @"\" + c.ToString());

            mwMessage = mwMessage.Replace("$1", "(?<item1>.+?)");
            mwMessage = mwMessage.Replace("$2", "(?<item2>.+?)");
            mwMessage = mwMessage.Replace("$3", "(?<item3>.+?)");
            mwMessage = mwMessage.Replace("$1", "(?:.+?)");
            mwMessage = mwMessage.Replace("$2", "(?:.+?)");
            mwMessage = mwMessage.Replace("$3", "(?:.+?)");
            mwMessage = mwMessage.Replace("$", @"\$");
            mwMessage = "^" + mwMessage + @"(?:: (?<comment>.*?))?$"; // Special:Log comments are preceded by a colon

            // Dirty code: Block log exceptions!
            if (mwMessageTitle == "Blocklogentry")
            {
                mwMessage = mwMessage.Replace("(?<item3>.+?)", "\\((?<item3>.+?)\\)");
                mwMessage = mwMessage.Replace(@"(?<item2>.+?)(?:: (?<comment>.*?))?$", "(?<item2>.+?)$");
            }

            try
            {
                Regex.Match("", mwMessage);
            }
            catch (Exception e)
            {
                throw new Exception("Failed to test-generate regex " + mwMessage + " for " + mwMessageTitle + "; " + e.Message);
            }

            if (reqCount >= 1)
            {
                if (!mwMessage.Contains(@"(?<item1>.+?)") && !nonStrict)
                    throw new Exception("Regex " + mwMessageTitle + " requires one or more items but item1 not found in "+mwMessage);
                if (reqCount >= 2)
                {
                    if (!mwMessage.Contains(@"(?<item2>.+?)") && !nonStrict)
                        throw new Exception("Regex " + mwMessageTitle + " requires two or more items but item2 not found in "+mwMessage);
                }
            }

            regexDict[destRegex] = mwMessage;
        }

        /// <summary>
        /// Gets the namespace code
        /// </summary>
        /// <param name="pageTitle">A page title, such as "Special:Helloworld" and "Helloworld"</param>
        /// <returns></returns>
        public int DetectNamespace(string pageTitle)
        {
            if (pageTitle.Contains(":"))
            {
                string nsLocal = pageTitle.Substring(0, pageTitle.IndexOf(':'));
                // Try to locate value (As fast as ContainsValue())
                foreach (DictionaryEntry de in this.namespaces)
                {
                    if ((string)de.Value == nsLocal)
                        return Convert.ToInt32(de.Key);
                }
            }
            // If no match for the prefix found, or if no colon,
            // assume main namespace
            return 0;
        }

        /// <summary>
        /// Returns a copy of the article title with the namespace translated into English
        /// </summary>
        /// <param name="originalTitle">Title in original (localized) language</param>
        /// <returns></returns>
        public static string TranslateNamespace(string project, string originalTitle)
        {
            if (originalTitle.Contains(":"))
            {
                string nsEnglish;

		        // *Don't change these* unless it's a stopping bug. These names are made part of the title
		        // in the watchlist and items database. (ie. don't change Image to File unless Image is broken)
		        // When they do need to be changed, make sure to make note in the RELEASE-NOTES that databases
		        // should be updated manually to keep all regexes and watchlists functional!
                switch (((Project)Program.prjlist[project]).DetectNamespace(originalTitle))
                {
                    case -2:
                        nsEnglish = "Media";
                        break;
                    case -1:
                        nsEnglish = "Special";
                        break;
                    case 1:
                        nsEnglish = "Talk";
                        break;
                    case 2:
                        nsEnglish = "User";
                        break;
                    case 3:
                        nsEnglish = "User talk";
                        break;
                    case 4:
                        nsEnglish = "Project";
                        break;
                    case 5:
                        nsEnglish = "Project talk";
                        break;
                    case 6:
                        nsEnglish = "Image";
                        break;
                    case 7:
                        nsEnglish = "Image talk";
                        break;
                    case 8:
                        nsEnglish = "MediaWiki";
                        break;
                    case 9:
                        nsEnglish = "MediaWiki talk";
                        break;
                    case 10:
                        nsEnglish = "Template";
                        break;
                    case 11:
                        nsEnglish = "Template talk";
                        break;
                    case 12:
                        nsEnglish = "Help";
                        break;
                    case 13:
                        nsEnglish = "Help talk";
                        break;
                    case 14:
                        nsEnglish = "Category";
                        break;
                    case 15:
                        nsEnglish = "Category talk";
                        break;
                    default:
                        return originalTitle;
                }

                // If we're still here, then nsEnglish has been set
                return nsEnglish + originalTitle.Substring(originalTitle.IndexOf(':'));
            }

			// Mainspace articles do not need translation
            return originalTitle;
        }
    }
}

File: cvnbot/project.py
import logging
import re
import xml.etree.ElementTree as ElementTree

from xml.sax.saxutils import escape

from . import utils

logger = logging.getLogger("CVNBot.Project")


class Project:
    """
    Namespaces and interface messages for one wiki.

    Used to parse and format a wiki's RC feed.

    Stored in Projects.xml, via projectlist.py.
    """

    # Characters that must be escaped when turning an interface message into a regex
    RE_CHARS = ["\\", ".", "(", ")", "[", "]", "^", "*", "+", "?", "{", "}", "|"]

    REGEX_DICT_KEYS = [
        "restoreRegex",
        "deleteRegex",
        "protectRegex",
        "unprotectRegex",
        "modifyprotectRegex",
        "uploadRegex",
        "moveRegex",
        "moveredirRegex",
        "blockRegex",
        "unblockRegex",
        "reblockRegex",
        "autosummBlank",
        "autosummReplace",
    ]

    def __init__(self):
        # Parsed in read_project_details(), stored in dump_project_details()
        self.project_name = ""
        self.interwiki_link = ""
        self.rooturl = ""  # Format: https://en.wikipedia.org/
        self.regex_dict = {}

        # Parsed via parse_namespaces()
        self.namespaces = {}
        # Parsed from regex_dict via generate_regexen()
        self.rrestore_regex = None
        self.rdelete_regex = None
        self.rprotect_regex = None
        self.runprotect_regex = None
        self.rmodifyprotect_regex = None
        self.rupload_regex = None
        self.rmove_regex = None
        self.rmoveredir_regex = None
        self.rblock_regex = None
        self.runblock_regex = None
        self.rreblock_regex = None
        self.rautosumm_blank = None
        self.rautosumm_replace = None
        self.rspecial_log_regex = None
        self.rcreate2_regex = None

    def generate_regexen(self):
        """
        Compile the regexen for this project.

        Always generate the namespace list before calling this!
        """
        self.rrestore_regex = utils.compile_dotnet(self.regex_dict["restoreRegex"])
        self.rdelete_regex = utils.compile_dotnet(self.regex_dict["deleteRegex"])
        self.rprotect_regex = utils.compile_dotnet(self.regex_dict["protectRegex"])
        self.runprotect_regex = utils.compile_dotnet(self.regex_dict["unprotectRegex"])

        if "modifyprotectRegex" not in self.regex_dict:
            # Added in CVNBot 1.20, fallback if missing in older XML files.
            self.regex_dict["modifyprotectRegex"] = self.regex_dict["protectRegex"]
            logger.warning(
                "generate_regexen: modifyprotectRegex is missing. Please reload this wiki."
            )
        self.rmodifyprotect_regex = utils.compile_dotnet(
            self.regex_dict["modifyprotectRegex"]
        )
        self.rupload_regex = utils.compile_dotnet(self.regex_dict["uploadRegex"])
        self.rmove_regex = utils.compile_dotnet(self.regex_dict["moveRegex"])
        self.rmoveredir_regex = utils.compile_dotnet(self.regex_dict["moveredirRegex"])
        self.rblock_regex = utils.compile_dotnet(self.regex_dict["blockRegex"])
        self.runblock_regex = utils.compile_dotnet(self.regex_dict["unblockRegex"])
        if "reblockRegex" not in self.regex_dict:
            # Added in CVNBot 1.22, fallback if missing in older XML files.
            self.regex_dict["reblockRegex"] = "^$"
            logger.warning(
                "generate_regexen: reblockRegex is missing. Please reload this wiki."
            )
        self.rreblock_regex = utils.compile_dotnet(self.regex_dict["reblockRegex"])
        self.rautosumm_blank = utils.compile_dotnet(self.regex_dict["autosummBlank"])
        self.rautosumm_replace = utils.compile_dotnet(self.regex_dict["autosummReplace"])

        self.regex_dict["specialLogRegex"] = self.namespaces["-1"] + r":.+?/(.+)"
        self.rspecial_log_regex = utils.compile_dotnet(self.regex_dict["specialLogRegex"])

        self.rcreate2_regex = re.compile(self.namespaces["2"] + r":([^:]+)")

    # -- Persistence ------------------------------------------------------

    def dump_project_details(self):
        """Serialize project details to a <project> element XML string."""
        parts = ["<project>"]

        def element(name, value):
            if value:
                parts.append("<{0}>{1}</{0}>".format(name, escape(value)))
            else:
                parts.append("<{0} />".format(name))

        element("projectName", self.project_name)
        element("interwikiLink", self.interwiki_link)
        element("rooturl", self.rooturl)
        element("speciallog", self.regex_dict["specialLogRegex"])

        namespaces_node = ElementTree.Element("namespaces")
        for key, value in self.namespaces.items():
            ns = ElementTree.SubElement(namespaces_node, "ns", id=key)
            ns.text = value
        element(
            "namespaces",
            ElementTree.tostring(namespaces_node, encoding="unicode"),
        )
        for name in self.REGEX_DICT_KEYS:
            element(name, self.regex_dict[name])

        parts.append("</project>")
        return "".join(parts)

    def read_project_details(self, element):
        """
        Load details from a <project> element from Projects.xml.

        Args:
            xml.etree.ElementTree.Element element
        """
        for child in element:
            value = child.text or ""
            if child.tag == "projectName":
                self.project_name = value
            if child.tag == "interwikiLink":
                self.interwiki_link = value
            if child.tag == "rooturl":
                self.rooturl = value
            elif child.tag == "speciallog":
                self.regex_dict["specialLogRegex"] = value
            elif child.tag == "namespaces":
                # Parse namespaces before generating regexen
                self.namespaces = Project.parse_namespaces(value)
            elif child.tag in self.REGEX_DICT_KEYS:
                self.regex_dict[child.tag] = value

        self.generate_regexen()

    @staticmethod
    def parse_namespaces(snamespaces):
        namespaces = {}
        namespaces_node = utils.parse_xml(snamespaces, "namespaces")
        if namespaces_node is None:
            raise Exception("No namespaces found")
        for child in namespaces_node:
            namespaces[child.get("id")] = child.text or ""
        return namespaces

    # -- Fetching from the wiki -------------------------------------------

    def get_namespaces(self):
        logger.info("Fetching namespaces from %s", self.rooturl)
        snamespaces = utils.get_raw_document(
            self.rooturl
            + "w/api.php?format=xml&action=query&meta=siteinfo&siprop=namespaces"
        )

        self.namespaces = Project.parse_namespaces(snamespaces)

    def retrieve_wiki_details(self):
        """
        Raises:
            Exception: if MediaWiki API request for namespaces fails.
        """

        # Find out what the localized Special: (ID -1) namespace is, and create a regex
        self.get_namespaces()

        logger.info("Fetching interface messages from %s", self.rooturl)

        self.get_interface_messages()

        self.generate_regexen()

    def get_interface_messages(self):
        # Interface message name -> (number of required parameters, regex_dict key, non-strict)
        INTERFACE_MESSAGES = {
            "Undeletedarticle": (1, "restoreRegex", False),
            "Deletedarticle": (1, "deleteRegex", False),
            "Protectedarticle": (1, "protectRegex", False),
            "Unprotectedarticle": (1, "unprotectRegex", False),
            "Modifiedarticleprotection": (1, "modifyprotectRegex", True),
            "Uploadedimage": (0, "uploadRegex", False),
            "1movedto2": (2, "moveRegex", False),
            "1movedto2_redir": (2, "moveredirRegex", False),

            # blockRegex is nonStrict because some wikis override the message without
            # including $2 (block length).
            # RCReader will fall back to "24 hours" if this is the case.
            # Some newer messages e.g. https://lmo.wikipedia.org/wiki/MediaWiki:Blocklogentry
            # have a third item $3 ("anononly,nocreate,autoblock"). This may conflict with
            # $2 detection.
            #
            # Trying (changed 2 -> 3) to see if length of time will be correctly detected
            # using just this method:
            "Blocklogentry": (3, "blockRegex", True),

            "Unblocklogentry": (0, "unblockRegex", False),
            "Reblock-logentry": (3, "reblockRegex", False),
            "Autosumm-blank": (0, "autosummBlank", False),

            # autosummReplace is nonStrict because some wikis use translation overrides
            # without a "$1" parameter for the content.
            "Autosumm-replace": (1, "autosummReplace", True),
        }
        raw = utils.get_raw_document(
            self.rooturl
            + "w/api.php?action=query&meta=allmessages&format=xml"
            + "&ammessages="
            + "|".join(INTERFACE_MESSAGES)
        )

        allmessages_node = utils.parse_xml(raw, "allmessages")
        if allmessages_node is None:
            raise Exception("No interface messages found for " + self.rooturl)
        for child in allmessages_node:
            name = child.get("name")
            required, dest_regex, non_strict = INTERFACE_MESSAGES[name]
            self.regex_dict[dest_regex] = self.generate_regex(
                name, child.text or "", required, non_strict
            )

    def generate_regex(self, message_title, message, req_count, non_strict):
        """Turn an interface message into a regex that matches itself."""

        # Now gently coax that into a regex
        for char in self.RE_CHARS:
            message = message.replace(char, "\\" + char)

        message = message.replace("$1", "(?<item1>.+?)", 1)
        message = message.replace("$2", "(?<item2>.+?)", 1)
        message = message.replace("$3", "(?<item3>.+?)", 1)
        message = message.replace("$1", "(?:.+?)")
        message = message.replace("$2", "(?:.+?)")
        message = message.replace("$3", "(?:.+?)")
        message = message.replace("$", r"\$")
        # Special:Log comments are preceded by a colon
        message = "^" + message + r"(?:: (?<comment>.*?))?$"

        # Dirty code: Block log exceptions!
        if message_title == "Blocklogentry":
            message = message.replace(
                "(?<item3>.+?)",
                "\\((?<item3>.+?)\\)"
            )
            message = message.replace(
                "(?<item2>.+?)(?:: (?<comment>.*?))?$",
                "(?<item2>.+?)$"
            )

        try:
            utils.compile_dotnet(message).search('')
        except re.error as e:
            raise Exception(
                "Failed to test-generate regex {0} for {1}; {2}".format(
                    message, message_title, e
                )
            )

        if req_count >= 1 and not non_strict:
            if "(?<item1>.+?)" not in message:
                raise Exception(
                    "Regex {0} requires one or more items but item1 not found in {1}".format(
                        message_title, message
                    )
                )
            if req_count >= 2 and "(?<item2>.+?)" not in message:
                raise Exception(
                    "Regex {0} requires two or more items but item2 not found in {1}".format(
                        message_title, message
                    )
                )

        return message

    # -- Namespaces -------------------------------------------------------

    def detect_namespace(self, page_title):
        """Get the namespace ID of a title such as "Special:Helloworld"."""
        if ":" in page_title:
            ns_local = page_title.split(":", 1)[0]
            for key, value in self.namespaces.items():
                if value == ns_local:
                    return int(key)
        # If no match for the prefix found, or if no colon,
        # assume main namespace
        return 0

    # DO NOT CHANGE OR REMOVE ANY VALUES. This is append-only.
    #
    # Incoming RC events from MediaWiki use titles formatted with the latest localised
    # namespace prefix.
    #
    # In ListManager, when adding/removing/showing titles in the "watchlist" or "items"
    # we noromalize the namespace prefixes to this canonical form so that titles reliably
    # match across wikis in different languages, and over time. For example, "Image"
    # changed to "File" many years ago, but this change must not be applied here.
    # If a change were needeed to fix a show-stopping bug, include release notes
    # and manually update existing databases to keep regexes and watchlists functional!
    #
    # In RCReader we also normalize titles to this form for the RCEvent message we
    # output to the IRC feed channels. This is mainly for multilingual channels like
    # #cvn-sw to help patrollers with unfamiliar namespace translations. The downside is
    # that monolingual channels like #cvn-wp-nl get titles presented in English among
    # an otherwise localised Console.msgs, and even #cvn-wp-en gets namespaces formatted
    # in an outdated form (File>Image).
    CANONICAL_NAMESPACES = {
        -2: "Media",
        -1: "Special",
        1: "Talk",
        2: "User",
        3: "User talk",
        4: "Project",
        5: "Project talk",
        6: "Image",
        7: "Image talk",
        8: "MediaWiki",
        9: "MediaWiki talk",
        10: "Template",
        11: "Template talk",
        12: "Help",
        13: "Help talk",
        14: "Category",
        15: "Category talk",
    }

    def translate_namespace(self, original_title):
        """Translate a local title's namespace to canonical English."""
        if ":" not in original_title:
            # Mainspace articles do not need translation
            return original_title

        ns_prefix = self.CANONICAL_NAMESPACES.get(self.detect_namespace(original_title))
        if ns_prefix is None:
            return original_title

        return ns_prefix + original_title[original_title.index(":"):]
